### ISSUE TITLE
Provide reasoning for `shared-lib-without-dependency-information` and `library-not-linked-against-libc`

### ISSUE THREAD

**Issue by voxik** [rpm-software-management/rpmlint#361]
It would be nice if the errors in subject provide some explanation why they are reported. Looking into git history, the intention behind these checks is not explained in the associated commits.

For example, checking ruby-libs-2.7.0-1.fc32.x86_64.rpm package, I get a lot of `library-not-linked-against-libc` errors but that is very likely just fine, because these libraries are utility libraries to Ruby and even if they use some glibc functionality, the glibc is already loaded.

Also, it would be nice to provide additional information how these issues could be analyzed. E.g. there could be suggested to use `readelf -d $(path/to/lib)` to obtain the similar information the rpmlint is checking.

**Comment by scarabeusiv** [comment]
Honestly no idea; this error is since 2001 and before that it seems to stretch to the begining as a warning.
Maybe @marxin has any idea what for this check would be? It might even be useless these days.

**Comment by junaruga** [comment]
I have an idea about `library-not-linked-against-libc`.

## Steps to reproduce the issues

```
$ rpm -q rpmlint
rpmlint-1.10-22.fc30.noarch
```

Download ruby-libs RPM from https://koji.fedoraproject.org/koji/taskinfo?taskID=40371136 .

```
$ wget https://kojipkgs.fedoraproject.org//work/tasks/1204/40371204/ruby-libs-2.7.0-1.fc32.x86_64.rpm
```

You see `library-not-linked-against-libc` error.

```
$ rpmlint ruby-libs-2.7.0-1.fc32.x86_64.rpm
...
ruby-libs.x86_64: E: library-not-linked-against-libc /usr/lib64/ruby/continuation.so
...
```

Then

```
$ rpm2cpio ruby-libs-2.7.0-1.fc32.x86_64.rpm | cpio -idmv

$ ls usr/lib64/ruby/continuation.so
usr/lib64/ruby/continuation.so*
```

```
$ ldd usr/lib64/ruby/continuation.so
  linux-vdso.so.1 (0x00007ffda938e000)
  libruby.so.2.7 => not found
```

Actually `libruby.so.2.7` exists in the same RPM.

```
$ ls -l usr/lib64/libruby.so.2.7
lrwxrwxrwx 1 jaruga jaruga 16 Jan 15 15:06 usr/lib64/libruby.so.2.7 -> libruby.so.2.7.0*
```

When using `LD_LIBRARY_PATH=./usr/lib64`, the libruby.so.2.7 file was found by `ldd`. and libc so files are found by it.

```
$ LD_LIBRARY_PATH=./usr/lib64 ldd usr/lib64/ruby/continuation.so
  linux-vdso.so.1 (0x00007ffc097dd000)
  libruby.so.2.7 => ./usr/lib64/libruby.so.2.7 (0x00007fbbc97f8000)
  libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fbbc97a2000)
  librt.so.1 => /lib64/librt.so.1 (0x00007fbbc9798000)
  libgmp.so.10 => /lib64/libgmp.so.10 (0x00007fbbc971b000)
  libdl.so.2 => /lib64/libdl.so.2 (0x00007fbbc9715000)
  libcrypt.so.2 => /lib64/libcrypt.so.2 (0x00007fbbc96da000)
  libm.so.6 => /lib64/libm.so.6 (0x00007fbbc9592000)
  libc.so.6 => /lib64/libc.so.6 (0x00007fbbc93cc000)
  /lib64/ld-linux-x86-64.so.2 (0x00007fbbc9b5b000)
```

So, I think `rpmlint` need to set `LD_LIBRARY_PATH` or something in the process of checking the RPM.

**Comment by junaruga** [comment]
For shared-lib-without-dependency-information,

## Steps to reproduce the issue: shared-lib-without-dependency-information

Using the same rpm-libs RPM.

```
$ rpmlint ruby-libs-2.7.0-1.fc32.x86_64.rpm
...
ruby-libs.x86_64: E: shared-lib-without-dependency-information /usr/lib64/ruby/enc/gb2312.so
...
```

```
$ rpm2cpio ruby-libs-2.7.0-1.fc32.x86_64.rpm | cpio -idmv
```

`statically linked` might be the cause of the error by `rpmlint`, and the `statically linked` state for the so file might be wrong.

```
$ ls usr/lib64/ruby/enc/gb2312.so
usr/lib64/ruby/enc/gb2312.so*

$ ldd usr/lib64/ruby/enc/gb2312.so
  statically linked
```

**Comment by junaruga** [comment]
For ruby-libs-2.6.5-124.fc32.x86_64.rpm, it is not `statically linked`.

```
$ ldd usr/lib64/ruby/enc/gb2312.so
	linux-vdso.so.1 (0x00007ffc15451000)
	libruby.so.2.6 => /lib64/libruby.so.2.6 (0x00007fece5a5c000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fece5a3b000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fece5a31000)
	libgmp.so.10 => /lib64/libgmp.so.10 (0x00007fece59b4000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fece59ae000)
	libcrypt.so.2 => /lib64/libcrypt.so.2 (0x00007fece5973000)
	libm.so.6 => /lib64/libm.so.6 (0x00007fece582b000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fece5665000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fece5dae000)
```

**Comment by marxin** [comment]
I incline to make warning our of these 2 errors.
About the provided RPM, I see very strange errors for `./usr/lib64/libruby.so.2.7.0`:

```
$ ldd ./usr/lib64/libruby.so.2.7.0 
ldd: warning: you do not have execution permission for `./usr/lib64/libruby.so.2.7.0'
ldd: exited with unknown exit code (135)
$ chmod +x ./usr/lib64/libruby.so.2.7.0
$ ldd ./usr/lib64/libruby.so.2.7.0 
ldd: exited with unknown exit code (135)
```

**Comment by junaruga** [comment]
Are you running the host CPU architecture's binary correctly like this?

```
$ uname -m
x86_64

$ file ./usr/lib64/libruby.so.2.7.0
./usr/lib64/libruby.so.2.7.0: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), dynamically linked, BuildID[sha1]=da7d05b901d3499d58142582992a508f241a3912, stripped
```

**Comment by marxin** [comment]
I see the following:
```
$ uname -m
x86_64
$ file ./usr/lib64/libruby.so.2.7.0 
./usr/lib64/libruby.so.2.7.0: ELF 64-bit LSB shared object, x86-64, version 1 (SYSV), statically linked, missing section headers
$ ~/Programming/bloaty/bloaty ./usr/lib64/libruby.so.2.7.0
bloaty: ELF region out-of-bounds
$ objdump -d ./usr/lib64/libruby.so.2.7.0 
objdump: ./usr/lib64/libruby.so.2.7.0: file truncated
$ ls -l ./usr/lib64/libruby.so.2.7.0
-rwx--x--x 1 marxin users 362496 Jan 15 16:25 ./usr/lib64/libruby.so.2.7.0
```

**Comment by marxin** [comment]
Sorry, my bad, it's for some reason caused by:
```
./usr/lib64/libruby.so.2.7
cpio: premature end of file
```

**Comment by marxin** [comment]
Now it's fine:
```
ldd ./usr/lib64/libruby.so.2.7.0
	linux-vdso.so.1 (0x00007ffc26d92000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007fbc6fce1000)
	librt.so.1 => /lib64/librt.so.1 (0x00007fbc6fcd6000)
	libgmp.so.10 => /usr/lib64/libgmp.so.10 (0x00007fbc6fc3f000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007fbc6fc3a000)
	libcrypt.so.2 => not found
	libm.so.6 => /lib64/libm.so.6 (0x00007fbc6faf5000)
	libc.so.6 => /lib64/libc.so.6 (0x00007fbc6f92c000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fbc7009f000)
```

**Comment by voxik** [comment]
> So, I think `rpmlint` needs to set `LD_LIBRARY_PATH` or something in the process of checking the RPM.

Unfortunately, this is wrong conclusion. Yes, ldd can't load the linked libraries, but that is not the problem, because rpmlint itself is using readelf, which is just reading the elf headers.

Apparently, typical .so file links gblic among their dependencies. I think that this is probably good thing in most cases. But since the Ruby libraries are just utility libraries used only by Ruby and nothing else, it is very likely OK that the libraries does not link got glibc.

@wenottingham since you are mentioned in b4d131907d0f4a1d299109c6af7512b776f08060 do you by a chance have explanation for this check at hand?

**Comment by junaruga** [comment]
Sorry I had not checked the rpmlint source code. My thought was just assumption.

**Comment by wenottingham** [comment]
That was a long time ago. I don't remember exactly, but if it uses any symbols from libc, it should have a DT_NEEDED for libc, certainly to ensure it uses the right symbol versions.

**Comment by voxik** [comment]
> That was a long time ago. I don't remember exactly, but if it uses any symbols from libc, it should have a DT_NEEDED for libc, certainly to ensure it uses the right symbol versions.

Thank you. That sounds as quite nice explanation for this test.

I don't necessarily think that this test has to go deep into analysis if there are any symbols from libc needed, but if there was explained also how to check that, that would help.

**Comment by marxin** [comment]
Based on the discussion, I tend to make `library-not-linked-against-libc` only warning as there can be multiple examples of such run-time libraries.
About the `shared-lib-without-dependency`, I really see it for the provided Fedora package. But openSUSE package looks fine for the file:
```
$ ldd /usr/lib64/ruby/2.6.0/x86_64-linux-gnu/enc/gb2312.so
	linux-vdso.so.1 (0x00007ffca79ef000)
	libruby2.6.so.2.6 => /usr/lib64/libruby2.6.so.2.6 (0x00007f3a41c6f000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f3a41aa8000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f3a41a86000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f3a41a7b000)
	libgmp.so.10 => /usr/lib64/libgmp.so.10 (0x00007f3a419e4000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f3a419df000)
	libcrypt.so.1 => /usr/lib64/libcrypt.so.1 (0x00007f3a419a2000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f3a4185d000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f3a41fc3000)
```
It may show a problem in package build?

**Comment by voxik** [comment]
This is interesting.

Ruby 2.7:

~~~
$ ldd /usr/lib64/ruby/enc/gb2312.so
	statically linked

$ readelf -d /usr/lib64/ruby/enc/gb2312.so

Dynamic section at offset 0x2e20 contains 22 entries:
  Tag        Type                         Name/Value
 0x000000000000000c (INIT)               0x1000
 0x000000000000000d (FINI)               0x1118
 0x0000000000000019 (INIT_ARRAY)         0x3e08
 0x000000000000001b (INIT_ARRAYSZ)       8 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x3e10
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x2f0
 0x0000000000000005 (STRTAB)             0x3a8
 0x0000000000000006 (SYMTAB)             0x318
 0x000000000000000a (STRSZ)              97 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x3fc0
 0x0000000000000002 (PLTRELSZ)           24 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x4b8
 0x0000000000000007 (RELA)               0x410
 0x0000000000000008 (RELASZ)             168 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x0000000000000018 (BIND_NOW)           
 0x000000006ffffffb (FLAGS_1)            Flags: NOW
 0x000000006ffffff9 (RELACOUNT)          3
 0x0000000000000000 (NULL)               0x0

$ rpm -q ruby-libs
ruby-libs-2.7.0-126.fc32.x86_64
~~~

Ruby 2.6:

~~~
$ ldd /usr/lib64/ruby/enc/gb2312.so
	linux-vdso.so.1 (0x00007fffb6b9e000)
	libruby.so.2.6 => /lib64/libruby.so.2.6 (0x00007f1a82b9d000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f1a82b7b000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f1a82b70000)
	libgmp.so.10 => /lib64/libgmp.so.10 (0x00007f1a82ad9000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f1a82ad2000)
	libcrypt.so.2 => /lib64/libcrypt.so.2 (0x00007f1a82a97000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f1a8294e000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f1a82785000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f1a82ed2000)

$ readelf -d /usr/lib64/ruby/enc/gb2312.so

Dynamic section at offset 0x2e00 contains 23 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libruby.so.2.6]
 0x000000000000000c (INIT)               0x1000
 0x000000000000000d (FINI)               0x1178
 0x0000000000000019 (INIT_ARRAY)         0x3de8
 0x000000000000001b (INIT_ARRAYSZ)       8 (bytes)
 0x000000000000001a (FINI_ARRAY)         0x3df0
 0x000000000000001c (FINI_ARRAYSZ)       8 (bytes)
 0x000000006ffffef5 (GNU_HASH)           0x2f0
 0x0000000000000005 (STRTAB)             0x3d8
 0x0000000000000006 (SYMTAB)             0x318
 0x000000000000000a (STRSZ)              140 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000003 (PLTGOT)             0x3fb0
 0x0000000000000002 (PLTRELSZ)           72 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x510
 0x0000000000000007 (RELA)               0x468
 0x0000000000000008 (RELASZ)             168 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x0000000000000018 (BIND_NOW)           
 0x000000006ffffffb (FLAGS_1)            Flags: NOW
 0x000000006ffffff9 (RELACOUNT)          3
 0x0000000000000000 (NULL)               0x0

$ rpm -q ruby-libs
ruby-libs-2.6.5-124.fc32.x86_64
~~~

**Comment by voxik** [comment]
Asking upstream:

https://bugs.ruby-lang.org/issues/16558

**Comment by scarabeusiv** [comment]
Ok the update for this bug is in the master branch now. I think we can close this, if you try it out and it still misbehaves feel free to open another tickets.

**Comment by yecril71pl** [comment]
> > That was a long time ago. I don't remember exactly, but if it uses any symbols from libc, it should have a DT_NEEDED for libc, certainly to ensure it uses the right symbol versions.
> 
> Thank you. That sounds as quite nice explanation for this test.

But it is not.  There are legitimate libraries that do not link against the C library.  I can think of two types: one is for number crunching and other specific maths, the other is object‑oriented CORBA‑style plug‑ins that use only callbacks provided by the host.

**Comment by yecril71pl** [comment]
> Ok the update for this bug is in the master branch now. I think we can close this, if you try it out and it still misbehaves feel free to open another tickets.

Should I open another ticket with the same summary?  The issue is closed even though no *sufficient reasoning* has been provided.

**Comment by marxin** [comment]
> Should I open another ticket with the same summary? The issue is closed even though no _sufficient reasoning_ has been provided.

I guess not.
Note that error became a warning and for the situation, you mentioned, it's likely easier to ignore the warnings in a `rpmlintrc`.
